### ISSUE TITLE
conftest.py fixtures do not behave the same way when packaged 

### ISSUE THREAD

**Issue by Swapnasheel** [pytest-dev/pytest#4168]
Hello,

I use pytest framework for testing and have generated a bunch of libraries functions for private use. 
I wish to distribute these libraries which also include the "conftest.py" files at each directory level.
Unfortunately, if I run tests that import fixtures do not behave the same as they do within my repo where I have these local fixtures.

I am using a fresh virtualenv while installing my package - that has its own dependencies. 

My directory structure looks something like this:

```
my_package:
|____ conftest.py
|____ directory_a
|........ |______ conftest.py
|____ directory_b
|........ |______ conftest.py
setup.py
requirements.py
```

When I package this, everything lives in the "/python2.7/site_packages/my_package" including the conftest.py files.

I want to make these fixtures visible in my package as the test scripts. So that I can even run my test cases simply by importing my_package.

Example script:

```python
import my_package

def test_connection(main_pc_connection):
    assert True
```

`main_pc_connection` is defined in the `my_package/conftest.py` file as a fixture with scope = "session"
This throws an error saying cannot find main_pc_connection.

**Comment by RonnyPfannschmidt** [comment]
its not quite clear what you want to do there from just the stuff you outline, but it seems to go completely against the structure eand expectations of pytest

please outline your actual use-case instead of your solution to it in order to avoid the XY-Problem

**Comment by nicoddemus** [comment]
Hi @Swapnasheel, 

As I understand it, you have a package `my_package` and a few fixtures, which right now are part of `my_package`'s test suite. Things have evolved, and you now need to use `my_package` in other projects, including the fixtures.

If that is the case, then all you need to do is to rename `conftest.py` to `fixtures.py` (or anything else other than `conftest.py` really), and [register that module as a pytest plugin](https://docs.pytest.org/en/latest/writing_plugins.html#making-your-plugin-installable-by-others) in your `setup.py`:

```python
setup(
    ...
    entry_points={"pytest11": ["my_package_plugin = my_package.fixtures"]},    
    classifiers=["Framework :: Pytest"],
)
```

As a plugin, its fixtures/hooks are now available both to `my_package`'s test suite and to other projects which install `my_package`.

If my understanding is incorrect, please state your use case in more detail as @RonnyPfannschmidt suggests.

(Also in the future, please format your code using Markdown so it is easier to read/follow, thanks!)

**Comment by Swapnasheel** [comment]
Right! @nicoddemus 
Now the project has grown and has a lot of libraries for various aspects of the project. 
These different aspects have their own `conftest.py` files in the respective directories.

Will I need to copy all of that conftest.py fixtures and add them at the `my_package/conftest.py` level (top most level), rename it anything but `conftest.py`, and add it as a plugin?

**Comment by nicoddemus** [comment]
> Will I need to copy all of that conftest.py fixtures and add them at the my_package/conftest.py level (top most level), rename it anything but conftest.py, and add it as a plugin?

The short answer is yes, this is the simplest approach.

The thing is that `conftest.py` files are local plugins, meaning that they are searched from the test module location and upwards in the filesystem. When your package is installed, it is installed into the virtual environment directory, which is usually not relative at all with the test modules of the projects which use `my_package`.

**Comment by Swapnasheel** [comment]
Alright!
Also, I use pytest-html plugins in the conftest.py file. I see that I can import the plugins with entry_points now but I cannot make use of the pytest-html features.

Is there a way to make that work as well?

**Comment by Swapnasheel** [comment]
`my_package` does include `pytest-html` in the `requirements.txt` file. 
But `pytest_configure` and related methods do not work for me as a plugin.

**Comment by Swapnasheel** [comment]
It does find the `pytest-HTML` when I run the script. But the execution doesn't find the `pytest_configure` method.

I also tried adding `@pytest.mark.optionalhook` marker on top of the `pytest_configure`, but still, the execution doesn't find the `pytest_configure` method in the `fixtures.py`.

Is there any way to make this work?

**Comment by nicoddemus** [comment]
> But pytest_configure and related methods do not work for me as a plugin.

You mean pytest does not call *your* `pytest_configure` function that used to reside in `conftest.py`?

> It does find the pytest-HTML when I run the script. 

Sorry, what do you mean by this? That you cannot import `pytest_html`, or something else?

It might be helpful if you post the exact error and snippets of your files as well.

**Comment by nicoddemus** [comment]
> but still it cannot run the def pytest_configure method, because I do not get my HTML generated anywhere.

What do you mean about the first part? Can you please elaborate? Also, can you please post the full output of adding `--trace-config` to the command-line running pytest (including the full command line)?

> Also, I have hard-coded the HTML output file path in pytest_configure. Is the hard-coding HTML output path a problem?

Shouldn't be...

**Comment by Swapnasheel** [comment]
Once my test script has run, it generates an `output.html` file - that is the default behavior if not packaged. From the above output, I know that `pytest-HTML` runs, but the `output.HTML` file is not getting created.

My output after `--trace-config`
`-------Starting test----------
============================================================================================= test session starts =============================================================================================
platform linux2 -- Python 2.7.12, pytest-3.7.1, py-1.5.4, pluggy-0.7.1 -- /home/swapnasheel/.virtualenvs/package_test/bin/python
cachedir: .pytest_cache
metadata: {'version': '1.0', 'tester': 'swapnasheel', 'Python': '2.7.12', 'Platform': 'Linux', 'Plugins': {'ordering': '0.5', 'metadata': '1.7.0', 'html': '1.19.0', 'my_package': '1.0.0'}, 'Packages': {'py': '1.5.4', 'pytest': '3.7.1', 'pluggy': '0.7.1'}}
rootdir: /home/swapnasheel/swapnasheel/test_package, inifile: pytest.ini
plugins: my_package-1.0.0, ordering-0.5, metadata-1.7.0, html-1.19.0
collected 1 item                                                                                                                                                                                              

new_test.py::test_something +++ Connection established to 127.0.0.1
INFO:paramiko.transport:Connected (version 2.0, client OpenSSH_7.2p2)
INFO:paramiko.transport:Authentication (publickey) failed.
INFO:paramiko.transport:Authentication (publickey) failed.
INFO:paramiko.transport:Authentication (password) successful!
INFO:paramiko.transport.sftp:[chan 0] Opened sftp connection (server version 3)
EXEC STATUS:mylogger:Testing requirement Swap123: Simple test
EXEC STATUS:mylogger:
Requirement Swap123 is being verified
Sample test
Requirement Swap123 PASSED: Expected: True, Actual: True

EXEC STATUS:mylogger:TEST PASSED
PASSED

============================================================================================== warnings summary ===============================================================================================
<undetermined location>
  Module already imported so cannot be rewritten: 

-- Docs: http://doc.pytest.org/en/latest/warnings.html
==================================================================================== 1 passed, 1 warnings in 6.29 seconds =====================================================================================`

**Comment by Swapnasheel** [comment]
`#pytest new_test.py -sv --trace-config

PLUGIN registered: <_pytest.config.PytestPluginManager object at 0x7f41ae3179d0>
PLUGIN registered: <_pytest.config.Config object at 0x7f41aaf691d0>
PLUGIN registered: <module '_pytest.mark' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/mark/__init__.pyc'>
PLUGIN registered: <module '_pytest.main' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/main.pyc'>
PLUGIN registered: <module '_pytest.terminal' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/terminal.pyc'>
PLUGIN registered: <module '_pytest.runner' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/runner.pyc'>
PLUGIN registered: <module '_pytest.python' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/python.pyc'>
PLUGIN registered: <module '_pytest.fixtures' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/fixtures.pyc'>
PLUGIN registered: <module '_pytest.debugging' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/debugging.pyc'>
PLUGIN registered: <module '_pytest.unittest' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/unittest.pyc'>
PLUGIN registered: <module '_pytest.capture' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/capture.pyc'>
PLUGIN registered: <module '_pytest.skipping' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/skipping.pyc'>
PLUGIN registered: <module '_pytest.tmpdir' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/tmpdir.pyc'>
PLUGIN registered: <module '_pytest.monkeypatch' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/monkeypatch.pyc'>
PLUGIN registered: <module '_pytest.recwarn' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/recwarn.pyc'>
PLUGIN registered: <module '_pytest.pastebin' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/pastebin.pyc'>
PLUGIN registered: <module '_pytest.helpconfig' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/helpconfig.pyc'>
PLUGIN registered: <module '_pytest.nose' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/nose.pyc'>
PLUGIN registered: <module '_pytest.assertion' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/assertion/__init__.pyc'>
PLUGIN registered: <module '_pytest.junitxml' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/junitxml.pyc'>
PLUGIN registered: <module '_pytest.resultlog' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/resultlog.pyc'>
PLUGIN registered: <module '_pytest.doctest' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/doctest.pyc'>
PLUGIN registered: <module '_pytest.cacheprovider' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/cacheprovider.pyc'>
PLUGIN registered: <module '_pytest.freeze_support' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/freeze_support.pyc'>
PLUGIN registered: <module '_pytest.setuponly' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/setuponly.pyc'>
PLUGIN registered: <module '_pytest.setupplan' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/setupplan.pyc'>
PLUGIN registered: <module '_pytest.warnings' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/warnings.pyc'>
PLUGIN registered: <module '_pytest.logging' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/logging.pyc'>
PLUGIN registered: <module 'my_package.fixtures' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/my_package/fixtures.py'>
PLUGIN registered: <module 'pytest_ordering' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/pytest_ordering/__init__.pyc'>
PLUGIN registered: <module 'pytest_metadata.plugin' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/pytest_metadata/plugin.py'>
PLUGIN registered: <module 'pytest_html.plugin' from '/home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/pytest_html/plugin.py'>
PLUGIN registered: <_pytest.capture.CaptureManager object at 0x7f41a12f2a90>
PLUGIN registered: <Session 'test_package'>
PLUGIN registered: <_pytest.cacheprovider.LFPlugin object at 0x7f41a1308f10>
PLUGIN registered: <_pytest.cacheprovider.NFPlugin object at 0x7f41a1310050>
PLUGIN registered: <_pytest.logging.LoggingPlugin object at 0x7f41a1310490>
PLUGIN registered: <_pytest.terminal.TerminalReporter object at 0x7f41a1310590>
PLUGIN registered: <_pytest.fixtures.FixtureManager object at 0x7f41a129d810>
============================================================================================= test session starts =============================================================================================
platform linux2 -- Python 2.7.12, pytest-3.7.1, py-1.5.4, pluggy-0.7.1 -- /home/swapnasheel/.virtualenvs/package/bin/python
using: pytest-3.7.1 pylib-1.5.4
setuptools registered plugins:
  my_package-1.0.0 at /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/my_package/fixtures.py
  pytest-ordering-0.5 at /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/pytest_ordering/__init__.pyc
  pytest-metadata-1.7.0 at /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/pytest_metadata/plugin.py
  pytest-html-1.19.0 at /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/pytest_html/plugin.py
active plugins:
    cacheprovider       : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/cacheprovider.pyc
    debugging           : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/debugging.pyc
    my_package_plugins      : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/my_package/fixtures.py
    pytestconfig        : <_pytest.config.Config object at 0x7f41aaf691d0>
    helpconfig          : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/helpconfig.pyc
    runner              : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/runner.pyc
    unittest            : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/unittest.pyc
    capturemanager      : <_pytest.capture.CaptureManager object at 0x7f41a12f2a90>
    pastebin            : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/pastebin.pyc
    skipping            : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/skipping.pyc
    session             : <Session 'test_package'>
    lfplugin            : <_pytest.cacheprovider.LFPlugin object at 0x7f41a1308f10>
    fixtures            : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/fixtures.pyc
    pytest_ordering     : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/pytest_ordering/__init__.pyc
    tmpdir              : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/tmpdir.pyc
    capture             : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/capture.pyc
    terminalreporter    : <_pytest.terminal.TerminalReporter object at 0x7f41a1310590>
    assertion           : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/assertion/__init__.pyc
    mark                : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/mark/__init__.pyc
    terminal            : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/terminal.pyc
    html                : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/pytest_html/plugin.py
    nfplugin            : <_pytest.cacheprovider.NFPlugin object at 0x7f41a1310050>
    main                : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/main.pyc
    nose                : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/nose.pyc
    warnings            : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/warnings.pyc
    python              : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/python.pyc
    recwarn             : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/recwarn.pyc
    funcmanage          : <_pytest.fixtures.FixtureManager object at 0x7f41a129d810>
    monkeypatch         : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/monkeypatch.pyc
    resultlog           : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/resultlog.pyc
    logging             : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/logging.pyc
    logging-plugin      : <_pytest.logging.LoggingPlugin object at 0x7f41a1310490>
    setuponly           : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/setuponly.pyc
    metadata            : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/pytest_metadata/plugin.py
    junitxml            : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/junitxml.pyc
    doctest             : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/doctest.pyc
    139920072079824     : <_pytest.config.PytestPluginManager object at 0x7f41ae3179d0>
    setupplan           : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/setupplan.pyc
    freeze_support      : /home/swapnasheel/.virtualenvs/package/local/lib/python2.7/site-packages/_pytest/freeze_support.pyc
cachedir: .pytest_cache
metadata: {'kiwi ver': '1.0', 'tester': 'swapnasheel', 'Python': '2.7.12', 'Platform': 'Linux-4.15.0-36-generic-x86_64-with-Ubuntu-16.04-xenial', 'Plugins': {'ordering': '0.5', 'metadata': '1.7.0', 'html': '1.19.0', 'my_package': '1.0.0'}, 'Packages': {'py': '1.5.4', 'pytest': '3.7.1', 'pluggy': '0.7.1'}}
rootdir: /home/swapnasheel/swapnasheel/test_package, inifile: pytest.ini
plugins: my_package-1.0.0, ordering-0.5, metadata-1.7.0, html-1.19.0
collected 1 item                                                                                                                                                                                              

new_test.py::test_something +++ Connection established to 127.0.0.1
INFO:paramiko.transport:Connected (version 2.0, client OpenSSH_7.2p2)
INFO:paramiko.transport:Authentication (publickey) failed.
INFO:paramiko.transport:Authentication (password) successful!
INFO:paramiko.transport.sftp:[chan 0] Opened sftp connection (server version 3)
EXEC STATUS:mylogger:Testing requirement Swap123: Simple test
EXEC STATUS:mylogger:
Requirement Swap123 is being verified
Sample test
Requirement Swap123 PASSED: Expected: True, Actual: True

EXEC STATUS:mylogger:TEST PASSED
PASSED

========================================================================================== 1 passed in 6.05 seconds ===========================================================================================`

**Comment by nicoddemus** [comment]
About, `pytest new_test.py -sv --trace-config`, don't you need to pass an option for pytest-html to generate the `.html` file?

**Comment by Swapnasheel** [comment]
My tests directory (without the installation of `my_package`) has a `pytest.ini` and a higher level `conftest.py` file. This higher level `conftest.py` file has `HTML` path hard coded.

I created a new `virtualenv` and installed `my_package` with the `fixture plugin` and executed the tests now with the same `pytest.ini` file. When I executed my tests with from the new `virtualenv` and with the `fixture plugin` (renaming the `conftest.py` as `fixtures.py`), I do not see the `output.html` generated (which happened by default in the different virtualenv without the `my_package`).

If I now give an argument `--html=output.html`, it ignores the hard-coded path of my `fixtures.py` plugin and stores `output.html` in my code execution directory. I only have confusion about this behavior for now. Because I did hard-code the `HTML` output path in `pytest_configure` method, but somehow that is being ignored.
